In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as py
import seaborn as sns
import category_encoders as ce
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import RandomForestClassifier

In [3]:
ps = pd.read_csv('C:\data\pulsarstars.csv')
#for colab
#ps = pd.read_csv('https://github.com/DoctorDroid/Build2/blob/master/pulsarstars.csv')
ps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17898 entries, 0 to 17897
Data columns (total 9 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0    Mean of the integrated profile                17898 non-null  float64
 1    Standard deviation of the integrated profile  17898 non-null  float64
 2    Excess kurtosis of the integrated profile     17898 non-null  float64
 3    Skewness of the integrated profile            17898 non-null  float64
 4    Mean of the DM-SNR curve                      17898 non-null  float64
 5    Standard deviation of the DM-SNR curve        17898 non-null  float64
 6    Excess kurtosis of the DM-SNR curve           17898 non-null  float64
 7    Skewness of the DM-SNR curve                  17898 non-null  float64
 8   target_class                                   17898 non-null  int64  
dtypes: float64(8), int64(1)
memory usage: 1.2 MB


## Clean. No null values and all of the data is numeric
### On to splitting the data into train/validate/and test sets

In [4]:
# Setting aside a saple of the data for testing.
# this portion will not be "known" during the model adjustment process.
train, test = train_test_split(ps,test_size=0.10, 
                              stratify=ps['target_class'],
                               random_state= 17)

In [5]:
# Separating the remaining data into training and validation sets.
#this allows me to tweak my model to perform at it's best before testing it on unknown data. 
train, val = train_test_split(train,test_size=0.20, 
                              stratify=train['target_class'],
                               random_state= 17)

In [6]:
train.shape,val.shape,test.shape

((12886, 9), (3222, 9), (1790, 9))

In [7]:
# To prove the similar distribuition over training, validation, and testing sets.
# This is accomplished by using the 'stratify' argument while splitting the data.
train['target_class'].value_counts(normalize=True),val['target_class'].value_counts(normalize=True),test['target_class'].value_counts(normalize=True)


(0    0.908428
 1    0.091572
 Name: target_class, dtype: float64,
 0    0.908442
 1    0.091558
 Name: target_class, dtype: float64,
 0    0.90838
 1    0.09162
 Name: target_class, dtype: float64)

## Solid.
###  If I just built a model that always reporterd '0' it would be correct about 90.8% or the time (our baseline accuracy). This may be dificult to improve upon, but I enjoy a challenge!
### Now to separate our data into X feature matrices and y target vectors.

In [8]:
target = 'target_class'
features = ps.columns.drop(target)

X_train = train[features]
y_train = train[target]
X_val = val[features]
y_val = val[target]
X_test = test[features]
y_test = test[target]

In [11]:
model = RandomForestClassifier(random_state=13,
                                  n_jobs=-1,
                                  n_estimators= 91,
                                  max_depth= 22)

# Fit on train, score on val
model.fit(X_train, y_train)
print('Validation Accuracy', model.score(X_val, y_val))

Validation Accuracy 0.9798261949099938


### After a few iterations, tweaking the hyperperameters lower and higher, I found the sweet spot of this models performance capabilities. Nearly 98% accurate is a better than expected result.
## Still I'll try for better before testing.
(I'd love to see another season of 'Lost in Space' , but I wouldn't want to be responsible for the reality-show version's release)